Added basic backtesting

j-waters opened the pull request:
It's rather slow and doesn't have any pretty logging - I'll get to that later.

I've created a mock Binance manager class, that simulates buying and selling, and can look up the historical ticker prices of coins. The API only allows you to get past data in 1 minute intervals, so the backtesting code runs the scout function in a loop without a delay, and after every scout an internal datetime value is incremented by 1 minute.

I also added some caching, <s>because as I said, it's rather slow. We can't do `get_all_market_tickers`, so each ticker has to be fetched individually, meaning that each run of the scout method takes about 3 seconds. It's not the end of the world, but it means testing a new config tweak isn't exactly instant. The cache does mean that once you've run the backtesting once, subsequent runs are a lot quicker. I added a script to fetch and cache data as quickly as possible, but then we run into Binance's throttling of 1200 requests per minute. Maybe we can sort something out where someone generates a sizeable amount of backtesting data and other people can just download it. Or maybe we get the data from a source other than Binance - it shouldn't make too much of a difference.</s>

j-waters commented:
Added a way of running backtesting

rel-s commented on README.md:
```diff
@@ -102,14 +104,30 @@ if you only want to start the sqlitebrowser
-## Developing
+## Backtesting
+
+You can test the bot on historic data to see how it performs
```
Formatting: period needed here.

rel-s replied:
Formatting: period needed here.

rel-s commented on README.md:
````diff
@@ -102,14 +104,30 @@ if you only want to start the sqlitebrowser
 docker-compose up -d sqlitebrowser
 ```
 
-## Developing
````
Why was this section removed? Seems unrelated to this PR.

rel-s reviewed (commented):
I love this idea - not only will it allow users to test the bot (and future strategies, configurations, etc) on historical data, it will also allow us to test code modifications in a way that makes a lot more sense. However, I looked over the changes files and it seems like quite the overhaul - what do you mean in that it can't be run without code modifications? Does this effect the "normal" user that does not expect this behaviour (other than adding the `trade` command)?

j-waters commented:
Ah yep there are a few rebasing issues sorry about that, I'll fix them later.

An early version of this PR didn't have a way of actually running the backtesting, which is why I said it didn't work without code modifications. That's no longer the case since I've added the `trade` and `backtest` commands.

My changes here don't affect a normal user whatsoever. Running the script like normal will default to running the `trade` command. 

The only other major change is the introduction of the `AllTickers` class. The reason I added this is because it makes it easier to mock ticker prices. I probably could have mocked this without any modifications to the rest of the code but it's so much cleaner and less convoluted this way. Also, I personally think the `AllTickers` class is a better way of doing things vs the tickers dictionary. If something like #205 is merged then this can be replaced.

j-waters commented:
Rebasing issues fixed.

Looking at this again I definitely think I could make a better system for running the backtesting, the CLI is a bit clunky. Maybe, since this is more of a technical feature, we simply have a script in the root of the project (not part of the package) that imports the package and mock manager. That script could be simple, clear and well documented so that anyone can easily set up their own tests. Personally, I'd probably want to run backtesting a few times tweaking different parameters, and that would be a lot easier to set up in a Python script vs a bunch of command line arguments

dederaditya commented:
I tried to run it, but instead of backtesting it tries to trade. Luckily, I put an API keys for reading only.

j-waters commented:
Sorry I forgot to update the README, I've changed how you run it. I've also fixed that issue

dederaditya commented:
```
2021-01-01 00:01:00.797093 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:02:00.907554 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:03:00.015360 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:04:00.122589 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:05:00.231528 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:06:00.342424 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:07:00.453290 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:08:00.581903 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:09:00.691293 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:10:00.799066 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:11:00.911891 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:12:00.101676 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:13:00.226173 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
2021-01-01 00:14:00.360547 - CONSOLE - INFO - I am scouting the best trades. Current coin: BTCUSDT
```

This is what I got after initializing, so it's working right?

j-waters commented:
I realised I was being a bit stupid before so have now managed to speed backtesting up massively! Before it took about 3 seconds per scout, now it takes about 0.5 seconds!

rel-s commented:
Ok, this is the next thing on my list. I'm not sure if I'll have time to do thorough testing today, but I'm on it.

rel-s commented:
First thing I noticed is that after running `backtest.py` I have a new untracked file, `.backtest_cache.db`. Perhaps we should add that to `.gitignore`?

dederaditya commented:
I got this error, it happened after I run the backtest for 5 minutes.

```
Traceback (most recent call last):
  File "backtest.py", line 7, in <module>
    for manager in backtest(datetime(2021, 3, 1), datetime.now()):
  File "/root/backtest/binance_trade_bot/backtest.py", line 176, in backtest
    trader.scout()
  File "/root/backtest/binance_trade_bot/auto_trader.py", line 152, in scout
    coin_opt_coin_ratio = current_coin_price / optional_coin_price
ZeroDivisionError: division by zero
```

dederaditya commented:
> I got this error, it happened after I run the backtest for 5 minutes. It happened while the bot stuck with a coin for days.
> 
> ```
> Traceback (most recent call last):
>   File "backtest.py", line 7, in <module>
>     for manager in backtest(datetime(2021, 3, 1), datetime.now()):
>   File "/root/backtest/binance_trade_bot/backtest.py", line 176, in backtest
>     trader.scout()
>   File "/root/backtest/binance_trade_bot/auto_trader.py", line 152, in scout
>     coin_opt_coin_ratio = current_coin_price / optional_coin_price
> ZeroDivisionError: division by zero
> ```

I think there was a maintenance on Binance system on 6/7 March? That's why the bot failed to backtest. It's working!

j-waters commented:
@dederaditya should be fixed

I also disabled apprise notifications while backtesting

bobwng commented:
I've ran the backtesting twice, both time got below error while run python backtest.py, looked into the code but currently no clue yet:

2021-03-22 20:23:45,124 - crypto_trading_logger - INFO - 2021-02-11 04:18:00
2021-03-22 20:23:45,429 - crypto_trading_logger - WARNING - Skipping increment: Couldn't get price for XMRUSDT
2021-03-22 20:23:45,429 - crypto_trading_logger - INFO - 2021-02-11 04:19:00
2021-03-22 20:23:45,689 - crypto_trading_logger - WARNING - Skipping increment: Couldn't get price for XMRUSDT
Traceback (most recent call last):
  File "C:\Users\Bob\git\binance-trade-bot\backtest.py", line 8, in <module>
    btc_value = manager.collate_coins("BTC")
  File "C:\Users\Bob\git\binance-trade-bot\binance_trade_bot\backtest.py", line 126, in collate_coins
    total += balance / self.get_market_ticker_price(target_symbol + coin)
  File "C:\Users\Bob\git\binance-trade-bot\binance_trade_bot\backtest.py", line 72, in get_market_ticker_price
    raise SkipIncrement(f"Couldn't get price for {ticker_symbol}")
binance_trade_bot.backtest.SkipIncrement: Couldn't get price for BTCUSDT

j-waters commented:
@bobwng Ah that would be because the pair `BTCUSDT` doesn't exist. I'll add something to catch that in the `collate_coins` method

mattangus commented on binance_trade_bot/auto_trader.py:
```diff
@@ -128,7 +127,7 @@ def scout(self):
         )
 
-        current_coin_price = get_market_ticker_price_from_list(all_tickers, current_coin + self.config.BRIDGE)
+        current_coin_price = all_tickers.get_price(current_coin + self.config.BRIDGE)
```
Might be good to add a `verbose` flag to the scout function to skip the print statement above when backtesting. The output I get is 80% `I am scouting the best trades. ...`

j-waters commented:
Also if #211 is merged the new `AllTickers` class won't be necessary

vaaski commented:
running `backtest.py` on my ubuntu VPS with default settings (apart from a readonly API key) it ran for a while but ultimately crashed.

i am not particularly familiar with python so i don't know what else info i to include, but i am happy to supply whatever you want to know about my setup.

<details>
<summary>traceback</summary>

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.8/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.8/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.8/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 328, in recv_into
    return self.recv_into(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 318, in recv_into
    raise SocketError(str(e))
OSError: (104, 'ECONNRESET')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 719, in urlopen
    retries = retries.increment(
  File "/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 400, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/lib/python3/dist-packages/six.py", line 702, in reraise
    raise value.with_traceback(tb)
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.8/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.8/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.8/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 328, in recv_into
    return self.recv_into(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 318, in recv_into
    raise SocketError(str(e))
urllib3.exceptions.ProtocolError: ('Connection aborted.', OSError("(104, 'ECONNRESET')"))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "backtest.py", line 7, in <module>
    for manager in backtest(datetime(2021, 1, 1), datetime.now()):
  File "/root/backtesting-binance/binance_trade_bot/backtest.py", line 187, in backtest
    trader.scout()
  File "/root/backtesting-binance/binance_trade_bot/auto_trader.py", line 141, in scout
    optional_coin_price = all_tickers.get_price(pair.to_coin + self.config.BRIDGE)
  File "/root/backtesting-binance/binance_trade_bot/backtest.py", line 21, in get_price
    return self.manager.get_market_ticker_price(ticker_symbol)
  File "/root/backtesting-binance/binance_trade_bot/backtest.py", line 63, in get_market_ticker_price
    for result in self.binance_client.get_historical_klines(
  File "/usr/local/lib/python3.8/dist-packages/binance/client.py", line 808, in get_historical_klines
    temp_data = self.get_klines(
  File "/usr/local/lib/python3.8/dist-packages/binance/client.py", line 734, in get_klines
    return self._get('klines', data=params, version=self.PRIVATE_API_VERSION)
  File "/usr/local/lib/python3.8/dist-packages/binance/client.py", line 237, in _get
    return self._request_api('get', path, signed, version, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/binance/client.py", line 202, in _request_api
    return self._request(method, uri, signed, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/binance/client.py", line 196, in _request
    self.response = getattr(self.session, method)(uri, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', OSError("(104, 'ECONNRESET')"))
```

</details>

j-waters commented:
@vaaski that looks like a connection issue to me, not caused by this PR but it is something that I can catch

ilearnio commented:
@j-waters I wonder what was your backtesting results? Does this strategy perform well?

moafred commented:
What is missing to merge this PR ?

bobwng commented on binance_trade_bot/backtest.py:
```diff
@@ -0,0 +1,204 @@
+    db.set_coins(config.SUPPORTED_COIN_LIST)
+    manager = MockBinanceManager(config, db, logger, start_date, start_balances)
+
+    starting_coin = db.get_coin(starting_coin or config.SUPPORTED_COIN_LIST[0])
+    if manager.get_currency_balance(starting_coin.symbol) == 0:
+        manager.buy_alt(starting_coin, config.BRIDGE, manager.get_all_market_tickers())
+    db.set_current_coin(starting_coin)
```
I'd like to suggest that put this part into strategy, the way to initialize the position should be part of the strategy rather than handled by backtest.

Imaging that we have another strategy which has different way to initialize the position. :)

bobwng replied:
Btw, this PR runs very well in my box, thank you very much, @j-waters !

bobwng commented on binance_trade_bot/backtest.py:
```diff
@@ -0,0 +1,204 @@
+        return FakeAllTickers(self)
+
+    def get_fee(self, origin_coin: Coin, target_coin: Coin, selling: bool):
+        return 0.0075
```
0.001 or 0.00075? ;)

77mena reviewed (commented):
Readme``

77mena commented on README.md:
```diff
@@ -92,7 +92,9 @@ If you are interested in running a Telegram bot, more information can be found a
 
 ### Run
```
Readme

77mena reviewed (commented):
Readme

77mena reviewed (commented):
Readme
